client/src/main/java/io/pravega/client/tables/TableWriter.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables;
+
+import com.google.common.annotations.Beta;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Defines all operations that can be used to modify a Table.
+ *
+ * Types of updates:
+ * * Unconditional Updates will insert and/or overwrite any existing values for the given Key, regardless of whether that Key
+ * previously existed or not, and regardless of what that Key's version is.
+ * * Conditional Updates will only overwrite an existing value if the specified {@link KeyVersion} matches that Key's version.
+ * * Unconditional Removals will remove a Key regardless of what that Key's version is. The operation will also succeed (albeit
+ * with no effect) if the Key does not exist.
+ * * Conditional Removals will remove a Key only if the specified {@link KeyVersion} matches that Key's version.
+ *
+ * @param <KeyT> Table Key Type.
+ * @param <ValueT> Table Value Type.
+ */
+@Beta
+public interface TableWriter<KeyT, ValueT> extends AutoCloseable {
+
+    /**
+     * Inserts a new Key-Value pair in this Table, but only if the Key does not already exist. Consider using
+     * {@link #put(KeyT, ValueT)} should the intent be to insert or update an existing key.
+     *
+     * @param key   The Key to insert or update.
+     * @param value The Value to set for the Key.
+     * @return A CompletableFuture that, when completed, will contain the {@link KeyVersion} associated with the newly
+     * inserted or updated entry.
+     */
+    CompletableFuture<KeyVersion> create(KeyT key, ValueT value);
+
+    /**
+     * Performs an Unconditional Update by inserting or updating the Value for a Key in this Table, regardless of whether
+     * the Key exists or not or what its Version is.
+     *
+     * @param key   The Key to insert or update.
+     * @param value The Value to set for the Key.
+     * @return A CompletableFuture that, when completed, will contain the {@link KeyVersion} associated with the newly
+     * inserted or updated entry.
+     */
+    CompletableFuture<KeyVersion> put(KeyT key, ValueT value);
+
+    /**
+     * Performs a Conditional Update by updating the Value for an existing Key in this Table.
```

Comment:
This is how most APIs work (look at ZK, BK, and pretty much all of our Tier 2 adapters) - if a conditional update fails, an exception is thrown at the caller, who must then resolve the problem or give up. The `ConditionalTableUpdateException` should contain sufficient information to resolve it, otherwise the caller must reconcile and retry.

I've updated the Javadoc.